### PR TITLE
feat(soap) :sparkles: create soap-client as class from unit_tests

### PR DESCRIPTION
**Pull request by dufrenb**

### CODE REVIEW

File: src/soap-client.service.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+      });
+    } catch (e: any) {
+      result = e.response;
+    }
+    this._soapClient.clearSoapHeaders();
+    return { messageUuid, result };
+  }
+
+  private _setClientSSLSecurityPFX(): void {
+    const pfx = fs.readFileSync('certificates/INSI-AUTO/AUTO-certificate.p12');
```
Comment (cadesalaberry) [inline src/soap-client.service.ts:89]: the pfx should be provided in the constructor

File: src/ins-soap-client.helper.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+export class INSSoapClientHelper {
+  public static getBAMContext(bamCtxId: string, { emitter }: SOAPINSConfig): object {
+    return {
+      ContexteBAM: {
+        attributes: {
+          Version: '01_02',
+        },
+        Id: bamCtxId,
+        Temps: moment().format(HEADER_DATE_FORMAT),
```
Comment (cadesalaberry) [inline src/ins-soap-client.helper.ts:14]: timestamp should be part of inputs

File: src/ins-soap-client.helper.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  public static getLPSContext(lpsCtxId: string, lpsCtxInstance: string, config: SOAPINSConfig): object {
+    return {
+      ContexteLPS: {
+        attributes: {
+          Nature: 'CTXLPS',
+          Version: '01_00',
+        },
+        Id: lpsCtxId,
+        Temps: moment().format(HEADER_DATE_FORMAT),
```
Comment (cadesalaberry) [inline src/ins-soap-client.helper.ts:29]: same here

File: src/ins-soap-client.service.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      ca: combineCACertAsPem([
+        'certificates/ca/ACR-EL.cer',
+        'certificates/ca/ACI-EL-ORG.cer',
+      ])
+    }))
+  }
+
+  private _setDefaultHeaders(): void {
+    this._soapClient.addSoapHeader(
+      INSSoapClientHelper.getBAMContext(this._bamCtxId, this._config),
```
Comment (cadesalaberry) [inline src/ins-soap-client.service.ts:102]: timestamps should be the same for all headers

File: src/ins-soap-client.service.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+  private readonly _bamCtxId: string;
+  private readonly _lpsCtxId: string;
+  private readonly _lpsCtxInstance: string;
+
+  private _soapClient: Client;
+
+  constructor(config: SOAPINSConfig) {
+    this._config = config;
+    this._bamCtxId = uuidv4();
```
Comment (cadesalaberry) [inline src/ins-soap-client.service.ts:52]: - How often do we need them to cycle?
- Does the lpsCtxId reflects the medical member's id in some way?
- With the instance being the employee id?

File: src/tryouts.test.ts [hunk 6]
```diff
@@ -63,93 +59,32 @@ describe('Using the p12 directly', () => {
...
+      IDAMAutorisationNumber: 'NumAuthorization'
+    })
+    await soapClient.initialize();
+  });
 
-    const MessageID = {
-      MessageID: '1f7425e2-b913-415c-adaa-785ee1076a70',
+  test('should error out when the software authorization number is unknown', async () => {
+    const fakePayload: INSIdentityTraits = {
+      lastName: 'ADRDEUX',
```
Comment (cadesalaberry) [inline src/tryouts.test.ts:79]: We are better of using the same naming to avoid confusion (NomNaissance...)

File: tsconfig.json [hunk 7]
```diff
@@ -81,7 +81,7 @@
     // "strictNullChecks": true,                         /* When type checking, take into account `null` and `undefined`. */
     // "strictFunctionTypes": true,                      /* When assigning functions, check to ensure parameters and the return values are subtype-compatible. */
     // "strictBindCallApply": true,                      /* Check that the arguments for `bind`, `call`, and `apply` methods match the original function. */
-    // "strictPropertyInitialization": true,             /* Check for class properties that are declared but not set in the constructor. */
+    "strictPropertyInitialization": false,             /* Check for class properties that are declared but not set in the constructor. */
```
Comment (cadesalaberry) [inline tsconfig.json:84]: weird. Which one is declared but never set?

File: src/ins-soap-client.service.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+      });
+    } catch (e: any) {
+      result = e.response;
+    }
+    this._soapClient.clearSoapHeaders();
+    return { messageUuid, result };
+  }
+
+  private _setClientSSLSecurityPFX(): void {
+    const pfx = fs.readFileSync('certificates/INSI-AUTO/AUTO-certificate.p12');
```
Comment (cadesalaberry) [inline src/ins-soap-client.service.ts:89]: pfx and passphrase should be in the constructor

File: tsconfig.json (hunk 7 above)
Comment (dufrenb) [reply tsconfig.json:84]: some, when we need to declare a variable before a try/catch and set it inside the other scope

File: src/ins-soap-client.service.ts (hunk 8 above)
Comment (dufrenb) [reply src/ins-soap-client.service.ts:89]: but not the certificates ?

File: src/ins-soap-client.service.ts (hunk 5 above)
Comment (dufrenb) [reply src/ins-soap-client.service.ts:52]: how would I know ? My job was just to implement SOAP client, not understand all the fields of the web service ^^

**Review by cadesalaberry — APPROVED** [review]

**Pull request merged**